### PR TITLE
Hide groups with new versions from old kubectl

### LINKED ISSUE #35791: Old kubectl crashes when v1alpha1 api is replaced with v1beta1.
```
kubectl get nodes -v 8

[...]
I1028 16:23:17.613562   26999 round_trippers.go:296] GET https://104.197.107.182/apis/policy/v1beta1
I1028 16:23:17.613645   26999 round_trippers.go:303] Request Headers:
I1028 16:23:17.613693   26999 round_trippers.go:306]     Accept: application/json, */*
I1028 16:23:17.613811   26999 round_trippers.go:306]     User-Agent: kubectl/v1.4.1 (linux/amd64) kubernetes/33cf7b9
I1028 16:23:17.613867   26999 round_trippers.go:306]     Authorization: Bearer bXlnXhEFV9oHeklOcVLX2Ri4IqFWArGx
I1028 16:23:17.727550   26999 round_trippers.go:321] Response Status: 200 OK in 113 milliseconds
I1028 16:23:17.727614   26999 round_trippers.go:324] Response Headers:
I1028 16:23:17.727646   26999 round_trippers.go:327]     Content-Length: 256
I1028 16:23:17.727676   26999 round_trippers.go:327]     Content-Type: application/json
I1028 16:23:17.727706   26999 round_trippers.go:327]     Date: Fri, 28 Oct 2016 14:23:17 GMT
I1028 16:23:17.727881   26999 request.go:908] Response Body: {"kind":"APIResourceList","apiVersion":"v1","groupVersion":"policy/v1beta1","resources":[{"name":"poddisruptionbudgets","namespaced":true,"kind":"PodDisruptionBudget"},{"name":"poddisruptionbudgets/status","namespaced":true,"kind":"PodDisruptionBudget"}]}
F1028 16:23:17.728389   26999 helpers.go:119] error: group map[:0xc8203d0540 authentication.k8s.io:0xc8203d0620 authorization.k8s.io:0xc8203d0770 autoscaling:0xc8203d07e0 batch:0xc8203d09a0 certificates.k8s.io:0xc8203d0a10 policy:0xc820070310 federation:0xc8203d0070 storage.k8s.io:0xc8200703f0 componentconfig:0xc8203d0a80 extensions:0xc8200701c0 rbac.authorization.k8s.io:0xc820070380 apps:0xc8203d05b0] is already registered
```

```
kubectl version
Client Version: version.Info{Major:"1", Minor:"4", GitVersion:"v1.4.1", GitCommit:"33cf7b9acbb2cb7c9c72a10d6636321fb180b159", GitTreeState:"clean", BuildDate:"2016-10-10T18:19:49Z", GoVersion:"go1.6.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"5+", GitVersion:"v1.5.0-alpha.1.985+fe0a941025aa36", GitCommit:"fe0a941025aa36ea48d37b6bec2e73fdca88185f", GitTreeState:"clean", BuildDate:"2016-10-28T11:34:34Z", GoVersion:"go1.7.1", Compiler:"gc", Platform:"linux/amd64"}
```

Is this expected? Does it mean that we cannot drop and replace alpha api without waiting for the new gcloud version?

### PR DESCRIPTION
**Pull request by caesarxuchao**
Fix https://github.com/kubernetes/kubernetes/issues/35791

**What caused the bug?**

In 1.5, we are going to graduate Policy and Apps to beta. Old version kubectl doesn't has the new versions built-in, its TRP dynamic discover thinks Policy/v1beta1 is a TPR, and tried to register it in kubectl's scheme. The registration failed because Policy group already exist, because kubectl had registered Policy.v1alpha1.

**How does this PR fix the bug?**

This PR let the API server hides Policy and Apps from old version kubectl, so TPR discovery won't see them.

Old version kubectl doesn't know about Policy/v1beta1 or Apps/v1beta1, and v1alpha1 will be removed, so old version kubectl won't work for Policy or Apps anyway, so this PR does not cause any function loss.

@kubernetes/sig-api-machinery @liggitt @smarterclayton @deads2k @janetkuo @mwielgus

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/35840)

### CODE REVIEW

**Comment by caesarxuchao** [comment]
@k8s-bot node e2e test this

File: pkg/genericapiserver/genericapiserver.go [hunk 1]
```diff
@@ -429,6 +430,43 @@ func (s *GenericAPIServer) newAPIGroupVersion(apiGroupInfo *APIGroupInfo, groupV
 	}, nil
 }
 
+// TODO: Remove in 1.6. Returns if kubectl is older than v1.5.0
+func isOldKubectl(userAgent string) bool {
```
Comment (liggitt) [inline pkg/genericapiserver/genericapiserver.go]: Not in genericapiserver, please. This only applies to the master, so make the surface area here as small as possible

**Comment by caesarxuchao** [comment]
Moved to apiserver. PTAL. Thanks.

File: pkg/apiserver/apiserver.go [hunk 2]
```diff
@@ -27,8 +27,11 @@ import (
 	"path"
 	rt "runtime"
 	"strconv"
+	"strings"
 	"time"
 
+	"k8s.io/client-go/pkg/util/sets"
+	"k8s.io/client-go/pkg/version"
```
Comment (liggitt) [inline pkg/apiserver/apiserver.go]: It's unfortunate goimports finds common types in the client-go pkg first. I think we want our own sets type

Comment (liggitt) [inline pkg/apiserver/apiserver.go]: Same for the version util

Comment (caesarxuchao) [inline pkg/apiserver/apiserver.go]: Eh..I'm using goimports locally and I guess it imports based on alphabetical order.

Comment (caesarxuchao) [inline pkg/apiserver/apiserver.go]: Fixed.

Comment (liggitt) [inline pkg/apiserver/apiserver.go]: I anticipate that creeping in a lot of places after client-go is added :-/

Comment (caesarxuchao) [inline pkg/apiserver/apiserver.go]: Yes. I think these kind of utility packages should be moved to their own repo, then there will be only one copy of them.

**Comment by janetkuo** [comment]
@liggitt @smarterclayton merging this PR will unblock two 1.5 PRs for moving StatefulSet to beta (#35731) and DisruptionBudget to beta (#35567)

File: pkg/apiserver/apiserver.go [hunk 3]
```diff
@@ -339,10 +342,48 @@ func APIVersionHandler(s runtime.NegotiatedSerializer, getAPIVersionsFunc func(r
...
+	kubectlVersion, versionErr := version.Parse(subs[1])
+	if versionErr != nil {
+		return false
+	}
+	return kubectlVersion.LT(version.MustParse("v1.5.0"))
+}
+
+// TODO: Remove in 1.6. This is for backward compatibility with 1.4 kubectl.
+// See https://github.com/kubernetes/kubernetes/issues/35791
+var groupsWithNewVersionsIn1_5 = sets.NewString("apps", "policy")
```
Comment (liggitt) [inline pkg/apiserver/apiserver.go]: both of these groups existed in 1.4... we're completely hiding the groups from the discovery doc?

**Comment by caesarxuchao** [comment]
@smarterclayton @liggitt could you take a look? Thanks.

**Comment by liggitt** [comment]
seems ok. can you queue up the revert to merge once the 1.5 branch is cut?

**Comment by caesarxuchao** [comment]
Do you mean revert in 1.6?

**Comment by liggitt** [comment]
yes

**Comment by caesarxuchao** [comment]
Right. Created #36004.

**Comment by caesarxuchao** [comment]
If there is no more comment I'll apply the label. It's blocking two 1.5 PRs.

**Comment by janetkuo** [comment]
> If there is no more comment I'll apply the label. It's blocking two 1.5 PRs.

Yes, presumably those PRs block other 1.5 PRs also. Would love to see this one merged ASAP.

**Comment by lavalamp** [comment]
IMO this is a client bug. I guess changing the server is an OK temporary fix, but the client should also change to tolerate collisions like this, so that we don't ever need any more server bandaids.

**Comment by caesarxuchao** [comment]
https://github.com/kubernetes/kubernetes/pull/36898 is the client-side fix. It's merged.

**Comment by lavalamp** [comment]
Oh, awesome!

**Comment by deads2k** [comment]
I agree.  I fixed it client-side shortly before release.

**Pull request merged**